core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -141,6 +146,32 @@ public String getDetails() {
         return this.details;
     }
 
+    /**
+     * The allocation delay value associated with the index (defaulting to node settings if not set).
+     */
+    public long getAllocationDelayTimeoutSetting(Settings settings, Settings indexSettings) {
+        if (reason != Reason.NODE_LEFT) {
+            return 0;
+        }
+        TimeValue delayTimeout = indexSettings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, settings.getAsTime(DELAYED_NODE_LEFT_TIMEOUT, DEFAULT_DELAYED_NODE_LEFT_TIMEOUT));
+        return Math.max(0l, delayTimeout.millis());
+    }
+
+    /**
+     * The delay in millis when delaying assigning the shard need to expire in.
+     */
+    public long getDelayAllocationExpirationIn(Settings settings, Settings indexSettings) {
+        long delayTimeout = getAllocationDelayTimeoutSetting(settings, indexSettings);
+        if (delayTimeout == 0) {
+            return 0;
+        }
+        long delta = Math.max(0l, System.currentTimeMillis() - timestamp);
+        if (delta == 0) {
```

Comment:
I think I see what you are doing here (protect against system clock going way back in time), but I wonder if we should be so strict.  0 is a dangerous value - I'm worried that a very fast machine will render this check ineffective, so I would go with `delta<0`. I also wonder if a short ntp correction will release pending shards. Maybe use `delta<-delayTimeout`? .